lib/src/default_index/composite.rs

```diff
@@ -514,9 +511,6 @@ impl<I: AsCompositeIndex> ChangeIdIndexImpl<I> {
 }
 
 impl<I: AsCompositeIndex + Send + Sync> ChangeIdIndex for ChangeIdIndexImpl<I> {
-    /// Resolves change id prefix among all ids, then filters out hidden
-    /// entries.
-    ///
```

Comment:
maybe removed the summary line by mistake?

Reply:
I moved this part of the documentation (with some tweaks) to the `ChangeIdIndex` trait. I left the rest because it seemed specific to this implementation. Maybe it should all be moved to the trait?

Reply:
I will move the rest of the comment over to the trait then. I think that the trait should probably "set the rules" about how the function works. Any documentation on the implementation should only cover implementation details that are relevant to the programmer. What do you think?

Reply:
Also, it's unclear why `AmbiguousMatch` would be returned even if the prefix is unique within visible entries. Do you know why that may be the case?

Reply:
Maybe turn it into non-doc comment, then? I expect inline comments are short and very specific to the implementation.

> Any documentation on the implementation should only cover implementation details that are relevant to the programmer.

I don't agree with that, but I won't insist if that's the convention.

> Also, it's unclear why `AmbiguousMatch` would be returned even if the prefix is unique within visible entries.

The search space is larger than `::visible_heads()`. As the comment says, this function has two steps:

1. resolve id among all (indexed) ids (including hidden entries)
2. filter out hidden

1 is stored on disk, 2 is in-memory data. `AmbiguousMatch` could be returned by 1 even if 2 could disambiguate them into a single entry.

Reply:
Ok thanks for clarifying. I understand how it works now but I think I misunderstood the function before. I updated the documentation and tried to make it more precise. Would you take another look please?

I'm not sure what the convention is... I'm just not sure it's a good idea to document how the trait should work, and then re-document the individual implementations. My concern is that the documentation can drift over time, or that the implementation will document that it doesn't conform to the documentation above the trait, which seems wrong.

In this case I just moved everything to the trait, if that's ok.

Reply:
> In this case I just moved everything to the trait, if that's ok.

Better to not do that. It's not the trait requirement, but an implementation detail of the default index.

The minimum requirement of the trait interface is something like:
* hidden revision shouldn't be included in the returned set
* should resolve to `SingleMatch` if visible id prefix longer than the `shortest_unique_prefix_len()` is provided.

Reply:
I agree that implementation-specific comments should be on the implementation and using `//` style comments (not `///` rustdoc style).

Reply:
Sorry Yuya for all of the back and forth, but I am having trouble seeing the difference between what you wrote and what I wrote.

It seems like the search space should be defined by the trait and be consistent across implementations, and it sounds like we are both saying that the search space is all change IDs in the index. Isn't that implicitly true if we say that `SingleMatch` is only returned if the visible id prefix is equal to or longer than `shortest_unique_prefix_len()` since we've defined `shortest_unique_prefix_len()` against every key in the index?

The docs for `shortest_unique_prefix_len` say:

https://github.com/martinvonz/jj/blob/fb5fe5cc26bc4fccc23bcb09860b14cc672199a8/lib/src/index.rs#L134-L135

Here is what I wrote for `resolve_prefix`. Would you mind pointing out the issues?

```
    /// Resolves a change ID prefix against all change IDs in the index, and
    /// returns:
    ///
    ///   - `NoMatch` if the prefix does not match any change IDs in the index, or if
    ///     all commits for the change ID it matches are hidden.
    ///   - `SingleMatch` if the prefix resolves unambiguously and points to one
    ///     or more visible commits. The order of the returned commits is
    ///     unspecified.
    ///   - `AmbiguousMatch` if the prefix matches multiple change IDs. Since
    ///     the search space is all change IDs in the index, `AmbiguousMatch`
    ///     may be returned even if the prefix is unambiguous against the set of
    ///     visible commits.
```

Reply:
> It seems like the search space should be defined by the trait and be consistent across implementations, and it sounds like we are both saying that the search space is all change IDs in the index.

Not really. The search space could be narrowed to `::visible_heads()` first if the index backend had a mechanism to compute that reasonably fast.

> Isn't that implicitly true if we say that `SingleMatch` is only returned if the visible id prefix is equal to or longer than `shortest_unique_prefix_len()` since we've defined `shortest_unique_prefix_len()` against every key in the index?

The default index implements `shortest_unique_prefix_len()` that way, but the trait requirement isn't. It's better if the index backend can compute the shortest length within `::visible_heads()`. The point is that, if `shortest_unique_prefix_len()` returns the true minimum length (within `::visible_heads()`), `resolve_prefix()` of the same index backend should be able to handle it consistently.

> The docs for `shortest_unique_prefix_len` say:
> 
> https://github.com/martinvonz/jj/blob/fb5fe5cc26bc4fccc23bcb09860b14cc672199a8/lib/src/index.rs#L134-L135

Maybe it's a bit ambiguous what "in the index" means. Historically, the change id index was built at runtime from `::visible_heads()`, so "the index" here was identical to `::visible_heads()` set. Conceptually, `ChangeIdIndex` is an index containing only `::visible_heads()` entries, but the backend is allowed to not compute the absolute minimum shortest unique prefix for performance reasons.

Reply:
Yeah, the magic is happening in the phrase "in the index", but maybe we can be more precise. It sounds like maybe what matters is that the prefix is disambiguated amongst the visible heads. However for performance reasons we relax the requirements for the trait such that `shortest_unique_prefix_len` doesn't have to actually return the shortest prefix length, as long as the length it returns disambiguates the change ID (and hopefully it is actually short too). 

But for `resolve_prefix` it's a little less clear how we should document this. It sounds like you're saying that the search space isn't all commits in the repository, but it also isn't restricted to visible heads in the default implementation. So what is it then? Maybe we should say that implementation are allowed to return `AmbiguousMatch` for performance reasons even if the prefix is unambiguous amongst the visible commits.

Anyways, I didn't expect this PR to evolve into such a big discussion, and I'm wary that it might be consuming too much of everyone's time. Let me know how you think we should document this and move forward.

Reply:
>the prefix is disambiguated amongst the visible heads. However for performance reasons we relax the requirements for the trait such that `shortest_unique_prefix_len` doesn't have to actually return the shortest prefix length, as long as the length it returns disambiguates the change ID (and hopefully it is actually short too).

Correct.

> But for `resolve_prefix` it's a little less clear how we should document this. It sounds like you're saying that the search space isn't all commits in the repository, but it also isn't restricted to visible heads in the default implementation. So what is it then? Maybe we should say that implementation are allowed to return `AmbiguousMatch` for performance reasons even if the prefix is unambiguous amongst the visible commits.
> 
> Anyways, I didn't expect this PR to evolve into such a big discussion, and I'm wary that it might be consuming too much of everyone's time. Let me know how you think we should document this and move forward.

I think the original trait doc (at some point in this PR) is good. I'm just against moving/removing the default implementation doc to the trait. So,
* restore the trait doc
* restore the implementation doc and turn it into non-doc comment?

Reply:
Ok, done. I still think that there is too much ambiguity in the bold part of this sentence:

"Resolve an unambiguous change ID prefix to **the commit IDs in the index**."

But let's move on for now :)